### PR TITLE
BLD: move metadata to setup.cfg

### PR DESCRIPTION
**Pull request by fangchenli**
move metadata to setup.cfg to separate them from build logic.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
 requires = [
     "setuptools",
     "wheel",
     "Cython>=0.29.21,<3",  # Note: sync with setup.py
-    "numpy==1.16.5; python_version=='3.7' and platform_system!='AIX'",
-    "numpy==1.17.3; python_version=='3.8' and platform_system!='AIX'",
-    "numpy==1.16.5; python_version=='3.7' and platform_system=='AIX'",
-    "numpy==1.17.3; python_version=='3.8' and platform_system=='AIX'",
+    "numpy==1.16.5; python_version=='3.7'",
+    "numpy==1.17.3; python_version=='3.8'",
```
Comment (fangchenli) [inline pyproject.toml:9]: copy scipy. min versions for AIX and non AIX are identical.

File: setup.cfg [hunk 2]
```diff
@@ -1,10 +1,61 @@
+[metadata]
+name = pandas
+description = Powerful data structures for data analysis, time series, and statistics
+long_description = file: README.md
+long_description_content_type = text/markdown
```
Comment (fangchenli) [inline setup.cfg:5]: If using README as the long description is not desired, I could revert this.

File: pyproject.toml [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
     "numpy; python_version>='3.9'",
 ]
+# uncomment to enable pep517 after versioneer problem is fixed.
+# https://github.com/python-versioneer/python-versioneer/issues/193
+# build-backend = "setuptools.build_meta"
```
Comment (fangchenli) [inline pyproject.toml:14]: We could modify PATH and uncomment this line to enable pep517. But I feel that may require some discussion.

**Comment by jreback** [comment]
great. can you build like this https://github.com/pandas-dev/pandas-release to make sure things look good (mainly testing that the sdist looks good).

File: setup.cfg (hunk 2 above)
Comment (jreback) [reply setup.cfg:5]: no that makes sense

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @simonjayhawkins if you give this a quick test as well. shall we backport to 1.2.1?

**Comment by fangchenli** [comment]
> great. can you build like this https://github.com/pandas-dev/pandas-release to make sure things look good (mainly testing that the sdist looks good).

Is uploading the sdist to testpypi the only way to test this? I couldn't get the version right. It won't allow me to upload the sdist.

**Comment by MarcoGorelli** [comment]
Nice! I'd suggesting adding https://github.com/asottile/setup-cfg-fmt as a pre-commit hook

**Comment by jreback** [comment]
@fangchenli you don't need to tag or upload 

just build the sdist in master and then on your PR and compare that they have the same metadata

File: setup.cfg [hunk 4]
```diff
@@ -61,47 +109,41 @@ xfail_strict = True
...
 
-    # Don't complain if tests don't hit defensive assertion code:
     raise AssertionError
     raise NotImplementedError
     AbstractMethodError
 
-    # Don't complain if non-runnable code isn't run:
     if 0:
     if __name__ == .__main__.:
     if TYPE_CHECKING:
```
Comment (MarcoGorelli) [inline setup.cfg:142]: Looks like we need to remove empty lines here for the linters to agree
```
exclude_lines =
    pragma: no cover
    def __repr__
    if self\.debug
    raise AssertionError
    raise NotImplementedError
    AbstractMethodError
    if 0:
    if __name__ == .__main__.:
    if TYPE_CHECKING:
```
setup-cfg-fmt` wants consistent indents, `remove-trailing-whitespace` doesn't want trailing whitespaces

File: setup.cfg [hunk 5]
```diff
@@ -1,13 +1,64 @@
...
+    pytest-xdist
 
 [build_ext]
-inplace = 1
+inplace = True
 
-# See the docstring in versioneer.py for instructions. Note that you must
-# re-run 'versioneer.py setup' after changing this section, and commit the
-# resulting files.
+[optioins.package_data]
```
Comment (simonjayhawkins) [inline setup.cfg:54]: typo

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -174,3 +174,7 @@ repos:
     -   id: codespell
         types_or: [python, rst, markdown]
         files: ^pandas/core/
+-   repo: https://github.com/asottile/setup-cfg-fmt
+    rev: v1.16.0
+    hooks:
+    -   id: setup-cfg-fmt
```
Comment (simonjayhawkins) [inline .pre-commit-config.yaml:180]: can you do this as a precursor

File: conda.recipe/meta.yaml [hunk 7]
```diff
@@ -19,7 +19,7 @@ requirements:
     - pip
     - cython
     - numpy
-    - setuptools >=3.3
+    - setuptools >=38.6.0
```
Comment (simonjayhawkins) [inline conda.recipe/meta.yaml:22]: this file is not used in the release process. I'm not sure if this has been maintained. Maybe worth removing.

**Comment by simonjayhawkins** [comment]
> cc @simonjayhawkins if you give this a quick test as well. shall we backport to 1.2.1?

pip-test and conda-test failures, 74 failed.

eg. `FileNotFoundError: [Errno 2] No such file or directory: '/opt/conda/envs/pip-test/lib/python3.7/site-packages/pandas/tests/io/parser/data/unicode_series.csv'`

**Comment by simonjayhawkins** [comment]
building from source and running `pandas.test(extra_args=['-m not clipboard', '--skip-slow', '--skip-network', '--skip-db', '-n=2'])` gives 87 failed which includes 13 failures seen at https://dev.azure.com/pandas-dev/pandas-wheels/_build/results?buildId=51595&view=logs&j=bb1c2637-64c6-57bd-9ea6-93823b2df951&t=9c62e7ff-391e-517d-ede1-af63bc6a25da&s=d654deb9-056d-50a2-1717-90c08683d50a

**Comment by jreback** [comment]
@simonjayhawkins these failures are in a 32 but build? are these not failing the ci now?

this PR is not related so this is concerning

**Comment by simonjayhawkins** [comment]
> this PR is not related so this is concerning

yeah, basically mentioning that if you get those 13 errors they are not from the changes here.

File: doc/source/getting_started/install.rst [hunk 8]
```diff
@@ -219,7 +219,7 @@ Dependencies
 ================================================================ ==========================
 Package                                                          Minimum supported version
 ================================================================ ==========================
-`setuptools <https://setuptools.readthedocs.io/en/latest/>`__    24.2.0
+`setuptools <https://setuptools.readthedocs.io/en/latest/>`__    38.6.0
```
Comment (simonjayhawkins) [inline doc/source/getting_started/install.rst:222]: from https://medium.com/@daveshawley/safely-using-setup-cfg-for-metadata-1babbe54c108 

> If you want to use configuration in setup.cfg, then you should make life a little better on your users and include the pkg_resources.require call in your setup.py. It means fewer confused users creating issues for you and makes even educate your users about one of the dirtier corners of python packaging.

**Comment by simonjayhawkins** [comment]
> shall we backport to 1.2.1?

@jreback if we are changing the minimum version for setuptools then maybe should leave this change till 1.3

@fangchenli the last time we changed the minimum supported version of setuptools (in pandas-0.23.0) we included the change in the release notes.

File: conda.recipe/meta.yaml (hunk 7 above)
Comment (jreback) [reply conda.recipe/meta.yaml:22]: this is needed for the conda recipe, which is itself not used, so yeah can prob remove (in its entirely), but can do later

**Comment by jreback** [comment]
> > shall we backport to 1.2.1?
> 
> @jreback if we are changing the minimum version for setuptools then maybe should leave this change till 1.3
> 
> @fangchenli the last time we changed the minimum supported version of setuptools (in pandas-0.23.0) we included the change in the release notes.

its fine to do this for 1.3

though our min on setuptools itself is pretty old.

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins these failures are in a 32 bit build? are these not failing the ci now?
> 
> this PR is not related so this is concerning

re-starting the job on MacPython (which pulls in master) and it passed on the 2nd attempt.

Merging master on this branch and those 13 tests now pass. (on linux with Python 3.8.5)

**Comment by simonjayhawkins** [comment]
> pip-test and conda-test failures, 74 failed.
> 
> eg. `FileNotFoundError: [Errno 2] No such file or directory: '/opt/conda/envs/pip-test/lib/python3.7/site-packages/pandas/tests/io/parser/data/unicode_series.csv'`

the tests that fail use the csv_dir_path fixture. AFAICT those tests should be failing since the directory exists (not all files are removed) but not the data file. so not sure why we have not seen these errors before now.

File: .pre-commit-config.yaml (hunk 6 above)
Comment (fangchenli) [reply .pre-commit-config.yaml:187]: I tried it. But this hook requires a [metadata] section.

File: doc/source/getting_started/install.rst (hunk 8 above)
Comment (fangchenli) [reply doc/source/getting_started/install.rst:222]: Or maybe set a min setuptools version in pyproject.toml?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@simonjayhawkins merge when ready

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins merge when ready

If my method of testing this was correct, I assume merging this now will break the nightly builds.

@fangchenli have you built from the sdist and got the same test failures?

**Comment by simonjayhawkins** [comment]
@fangchenli I've added an extra ci workflow that can be removed once the tests pass.

The tests failing are those using the `csv_dir_path` fixture. There are also a couple of extra failures to those I've seen locally due to version number issues that can be ignored for now.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -132,3 +132,50 @@ jobs:
...
+        sudo apt-get clean
+    - name: Setting conda path
+      run: echo "$CONDA/bin" >> $GITHUB_PATH
+    - name: Update conda
+      run: |
+        conda config --set quiet true --set always_yes true
+        conda update -n base -c defaults conda
+        conda list
+    - name: Checkout pandas
+      uses: actions/checkout@v2
```
Comment (fangchenli) [inline .github/workflows/ci.yml:153]: I'm not sure if it's related here. But using checkout@v2 caused a lot of failures and errors in #38344.

Comment (simonjayhawkins) [reply .github/workflows/ci.yml:153]: maybe the cause of the version issues. I added the step to pull the tags but that didn't seem to fix.

I don't have the issue in #36704, but am adding tag explicitly there and using fetch-depth with the checkout@v2 action.

File: setup.cfg [hunk 10]
```diff
@@ -38,24 +89,19 @@ bootstrap =
...
+    E402,  # module level import not at top of file
+    W503,  # line break before binary operator
+    E302,  # expected 2 blank lines, found 0
+    E305,  # expected 2 blank lines after class or function definition, found 0
+    E703,  # statement ends with a semicolon
+    E711,  # comparison to none should be 'if cond is none:'
 exclude =
     doc/source/development/contributing_docstring.rst,
-    # work around issue of undefined variable warnings
-    # https://github.com/pandas-dev/pandas/pull/38837#issuecomment-752884156
```
Comment (jorisvandenbossche) [inline setup.cfg:54]: Is there a reason you are removing those comments? (and below some occurrences as well)

File: setup.cfg [hunk 11]
```diff
@@ -1,13 +1,64 @@
+[metadata]
+name = pandas
+description = Powerful data structures for data analysis, time series, and statistics
+long_description = file: README.md
+long_description_content_type = text/markdown
+url = https://pandas.pydata.org
+author = The PyData Development Team
+author_email = pydata@googlegroups.com
```
Comment (jorisvandenbossche) [inline setup.cfg:8]: While at it, should probably update this to pandas-dev@python.org, and the "author"

File: setup.cfg (hunk 10 above)
Comment (fangchenli) [reply setup.cfg:54]: Maybe we shouldn't use setup-cfg-fmt.

Comment (jorisvandenbossche) [reply setup.cfg:54]: If there is no way to configure the formatter to keep the comments, then I think we should indeed not use it

File: setup.py [hunk 12]
```diff
@@ -711,51 +620,11 @@ def srcpath(name=None, suffix=".pyx", subdir="src"):
...
 
 
-def setup_package():
-    setuptools_kwargs = {
-        "install_requires": [
-            "python-dateutil >= 2.7.3",
-            "pytz >= 2017.3",
-            f"numpy >= {min_numpy_ver}",
-        ],
-        "setup_requires": [f"numpy >= {min_numpy_ver}"],
```
Comment (jorisvandenbossche) [inline setup.py:721]: It seems `setup_requires` wasn't added back in setup.cfg? Are there consequences of no longer including it?

File: setup.cfg (hunk 10 above)
Comment (jorisvandenbossche) [reply setup.cfg:54]: Apparently not implemented at the moment, see https://github.com/asottile/setup-cfg-fmt/pull/21#issuecomment-533308788. 
So I would propose to remove the use of setup-cfg-fmt, and add back the comments (can of course keep the other formatting improvements)

File: setup.py (hunk 12 above)
Comment (simonjayhawkins) [reply setup.py:721]: The added workflow pip installs pandas without the requirements pre-installed. Seems to pull in requirements OK. (I also wanted to test with other versions of setuptools locally though before approving this PR)

OT for the sdist build step we do `python setup.py clean --quiet`. This fails if numpy not installed.

Comment (jorisvandenbossche) [reply setup.py:721]: But you installed numpy in that workflow, so the question is maybe what would happen if numpy was not installed? (having the setup_requires might give a nicer error message?)

Comment (simonjayhawkins) [reply setup.py:721]: That's why I created 2 different conda environments. so numpy is not in the pip-test environment. (In  #36704 the build and test steps are done independently - here I just added a stripped down workflow that we maybe could consider adding permanently to the the ci)

**Comment by simonjayhawkins** [comment]
@fangchenli you have removed the install and test from the source dist?

**Comment by fangchenli** [comment]
> @fangchenli you have removed the install and test from the source dist?

I accidentally removed it while merging master... Sorry about that. I've added it back.

**Comment by simonjayhawkins** [comment]
> > @fangchenli you have removed the install and test from the source dist?
> 
> I accidentally removed it while merging master... Sorry about that. I've added it back.

no problem. can remove before merge but need to sort out why the tests are failing with these changes and not before when pip installing sdist directly

**Comment by fangchenli** [comment]
> > > @fangchenli you have removed the install and test from the source dist?
> > 
> > 
> > I accidentally removed it while merging master... Sorry about that. I've added it back.
> 
> no problem. can remove before merge but need to sort out why the tests are failing with these changes and not before when pip installing sdist directly

Most of those errors are FileNotFound. Does it mean the sdist doesn't include those data files?

**Comment by simonjayhawkins** [comment]
> Most of those errors are FileNotFound. Does it mean the sdist doesn't include those data files?

no they don't, but AFAIK never have.

the difference maybe in other files being included and so the directory exists any maybe it didn't before.

the issue is actually with the csv_dir_path fixture which incorrectly checks for the presence of the directory and not the file

File: MANIFEST.in [hunk 13]
```diff
@@ -1,9 +1,4 @@
-include MANIFEST.in
-include LICENSE
 include RELEASE.md
-include README.md
-include setup.py
-include pyproject.toml
```
Comment (fangchenli) [inline MANIFEST.in:6]: According to https://packaging.python.org/guides/using-manifest-in/, those are included by default.

**Comment by fangchenli** [comment]
> the issue is actually with the csv_dir_path fixture which incorrectly checks for the presence of the directory and not the file

I excluded that dir in MANIFEST.in. It's working now.

**Comment by simonjayhawkins** [comment]
> > the issue is actually with the csv_dir_path fixture which incorrectly checks for the presence of the directory and not the file
> 
> I excluded that dir in MANIFEST.in. It's working now.

Thanks. You can remove that extra ci job now.

File: MANIFEST.in [hunk 14]
```diff
@@ -40,6 +37,8 @@ global-exclude .DS_Store
 global-exclude .git*
 global-exclude \#*
 
+prune pandas/tests/io/parser/data
```
Comment (simonjayhawkins) [inline MANIFEST.in:40]: maybe add a comment here why this is excluded. could maybe open an issue about csv_dir_path fixture and reference it here.

**Comment by jreback** [comment]
lgtm. @simonjayhawkins

File: MANIFEST.in [hunk 15]
```diff
@@ -40,6 +37,9 @@ global-exclude .DS_Store
 global-exclude .git*
 global-exclude \#*
 
+# GH 39321
```
Comment (jorisvandenbossche) [inline MANIFEST.in:40]: Could you include a brief explainer as well, instead of only the issue reference?

File: setup.cfg [hunk 16]
```diff
@@ -38,16 +92,14 @@ bootstrap =
     import pandas as pd
     np  # avoiding error when importing again numpy or pandas
     pd  # (in some cases we want to do it to show users)
-ignore = E203,  # space before : (needed for how black formats slicing)
-         E402,  # module level import not at top of file
-         W503,  # line break before binary operator
-         # Classes/functions in different blocks can generate those errors
```
Comment (jorisvandenbossche) [inline setup.cfg:44]: this comment is still lost

File: setup.cfg [hunk 17]
```diff
@@ -38,16 +92,14 @@ bootstrap =
...
-         E302,  # expected 2 blank lines, found 0
-         E305,  # expected 2 blank lines after class or function definition, found 0
-         # We use semicolon at the end to avoid displaying plot objects
```
Comment (jorisvandenbossche) [inline setup.cfg:47]: and this one as well

**Review by jorisvandenbossche — COMMENTED** [review]
> You can remove that extra ci job now.

@simonjayhawkins should we consider keeping it? (I don't know how useful it is, or eg only on master)

**Comment by jreback** [comment]
@simonjayhawkins @jorisvandenbossche this lgtm. looks like final comments have been addressed. merging tomorrow.

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins should we consider keeping it? (I don't know how useful it is, or eg only on master)

we should maybe have this or something similar on ci. At the moment we will catch build errors on the nightly build.

but lets keep that discussion out of this PR.

**Comment by jorisvandenbossche** [comment]
OK, sounds good

**Comment by simonjayhawkins** [comment]
@fangchenli maybe we should open issues for conditional numpy import https://github.com/pandas-dev/pandas/pull/38852#discussion_r552479485 and ci check https://github.com/pandas-dev/pandas/pull/38852#discussion_r552490251

**Comment by jreback** [comment]
@fangchenli one more rebase and then will merge. ping on green.

**Comment by simonjayhawkins** [comment]
Thanks @fangchenli

**Comment by simonjayhawkins** [comment]
@fangchenli the wheels doubled in size around the same time that this was merged. Could it be this PR that is responsible? https://anaconda.org/scipy-wheels-nightly/pandas/files?page=32

**Comment by fangchenli** [comment]
> @fangchenli the wheels doubled in size around the same time that this was merged. Could it be this PR that is responsible? https://anaconda.org/scipy-wheels-nightly/pandas/files?page=32

Looking into it.

**Pull request merged**